File: src/ru/hudspeed/pro/StrelkaAlertTracker.java
package ru.gpsantiradar.app;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

/** Stateful approach confirmation and re-alert behavior reconstructed from Strelka HUD. */
public final class StrelkaAlertTracker {
    private final Map<Long, State> states = new HashMap<>();

    public Update update(List<Observation> observations, float speedKmh,
                         boolean gpsRecovered) {
        for (State state : states.values()) state.observed = false;

        for (Observation observation : observations) {
            State state = states.get(observation.object.id);
            boolean created = false;
            if (state == null && observation.matchesZone) {
                state = new State(observation.object);
                states.put(observation.object.id, state);
                created = true;
            }
            if (state == null) continue;

            state.observed = true;
            state.distanceMeters = observation.distanceMeters;
            state.activationDistance = observation.activationDistance;
            if (observation.dropImmediately) {
                state.confidence = 0f;
            } else {
                state.confidence = StrelkaAlertAlgorithm.updateConfidence(
                        state.confidence, speedKmh, observation.matchesZone,
                        gpsRecovered && created);
            }
            if (!state.active && StrelkaAlertAlgorithm.shouldActivate(
                    state.confidence, state.distanceMeters)) {
                state.active = true;
            }
        }

        List<State> exited = new ArrayList<>();
        Iterator<State> iterator = states.values().iterator();
        while (iterator.hasNext()) {
            State state = iterator.next();
            if (!state.observed || state.confidence <= 0f) {
                if (state.active) exited.add(state);
                iterator.remove();
            }
        }

        return summarize(exited);
    }

    public Update snapshot() {
        return summarize(new ArrayList<State>());
    }

    private Update summarize(List<State> exited) {
        State closestActive = null;
        State pendingVoice = null;
        State closestTracking = null;
        int activeCount = 0;
        for (State state : states.values()) {
            if (closestTracking == null
                    || state.distanceMeters < closestTracking.distanceMeters) {
                closestTracking = state;
            }
            if (!state.active) continue;
            activeCount++;
            if (closestActive == null
                    || state.distanceMeters < closestActive.distanceMeters) {
                closestActive = state;
            }
            if (!state.spoken && (pendingVoice == null
                    || state.distanceMeters < pendingVoice.distanceMeters)) {
                pendingVoice = state;
            }
        }
        return new Update(exited, closestActive, pendingVoice,
                closestTracking, activeCount);
    }

    public void markSpoken(long objectId) {
        State state = states.get(objectId);
        if (state != null) state.spoken = true;
    }

    public void clear() {
        states.clear();
    }

    public static final class Observation {
        final CameraPoint object;
        final int distanceMeters;
        final int activationDistance;
        final boolean matchesZone;
        final boolean dropImmediately;

        public Observation(CameraPoint object, int distanceMeters,
                           int activationDistance, boolean matchesZone,
                           boolean dropImmediately) {
            this.object = object;
            this.distanceMeters = distanceMeters;
            this.activationDistance = activationDistance;
            this.matchesZone = matchesZone;
            this.dropImmediately = dropImmediately;
        }
    }

    public static final class State {
        final CameraPoint object;
        float confidence;
        boolean active;
        boolean spoken;
        boolean observed;
        int distanceMeters;
        int activationDistance;

        State(CameraPoint object) {
            this.object = object;
        }
    }

    public static final class Update {
        final List<State> exited;
        final State closestActive;
        final State pendingVoice;
        final State closestTracking;
        final int activeCount;

        Update(List<State> exited, State closestActive, State pendingVoice,
               State closestTracking, int activeCount) {
            this.exited = exited;
            this.closestActive = closestActive;
            this.pendingVoice = pendingVoice;
            this.closestTracking = closestTracking;
            this.activeCount = activeCount;
        }
    }
}

File: src/main/kotlin/ru/gpsantiradar/app/StrelkaAlertTracker.kt
package ru.gpsantiradar.app

/** Stateful approach confirmation and re-alert behavior reconstructed from Strelka HUD. */
class StrelkaAlertTracker {
    private val states = HashMap<Long, State>()

    fun update(observations: List<Observation>, speedKmh: Float, gpsRecovered: Boolean): Update {
        states.values.forEach { it.observed = false }
        for (observation in observations) {
            var state = states[observation.`object`.id]
            var created = false
            if (state == null && observation.matchesZone) {
                state = State(observation.`object`)
                states[observation.`object`.id] = state
                created = true
            }
            if (state == null) continue
            state.observed = true
            state.distanceMeters = observation.distanceMeters
            state.activationDistance = observation.activationDistance
            state.confidence = if (observation.dropImmediately) 0f else StrelkaAlertAlgorithm.updateConfidence(
                state.confidence, speedKmh, observation.matchesZone, gpsRecovered && created
            )
            if (!state.active && StrelkaAlertAlgorithm.shouldActivate(state.confidence, state.distanceMeters)) state.active = true
        }
        val exited = ArrayList<State>()
        val iterator = states.values.iterator()
        while (iterator.hasNext()) {
            val state = iterator.next()
            if (!state.observed || state.confidence <= 0f) {
                if (state.active) exited += state
                iterator.remove()
            }
        }
        return summarize(exited)
    }

    fun snapshot() = summarize(ArrayList())
    private fun summarize(exited: MutableList<State>): Update {
        val ordered = states.values.sortedWith(nearestFirst)
        exited.sortWith(nearestFirst)
        var closestActive: State? = null
        var pendingVoice: State? = null
        val closestTracking = ordered.firstOrNull()
        val active = ArrayList<State>()
        for (state in ordered) {
            if (!state.active) continue
            active += state
            if (closestActive == null) closestActive = state
            if (!state.spoken && pendingVoice == null) pendingVoice = state
        }
        return Update(exited, closestActive, pendingVoice, closestTracking, active)
    }

    fun markSpoken(objectId: Long) { states[objectId]?.spoken = true }
    fun clear() = states.clear()

    class Observation(
        val `object`: CameraPoint,
        val distanceMeters: Int,
        val activationDistance: Int,
        val matchesZone: Boolean,
        val dropImmediately: Boolean
    )

    class State internal constructor(val `object`: CameraPoint) {
        var confidence = 0f
        var active = false
        var spoken = false
        var observed = false
        var distanceMeters = 0
        var activationDistance = 0
    }

    class Update internal constructor(
        exited: List<State>,
        val closestActive: State?,
        val pendingVoice: State?,
        val closestTracking: State?,
        active: List<State>
    ) {
        val exited = exited.toList()
        private val active = active.toList()
        val activeCount = active.size
        val activeCameraIds: LongArray = active.map { it.`object`.id }.toLongArray()
        fun overspeedCandidate(speedKmh: Float) = overspeedCandidate(speedKmh, AppSettings.DEFAULT_OVERSPEED_THRESHOLD_KMH)
        fun overspeedCandidate(speedKmh: Float, thresholdKmh: Int): State? =
            active.firstOrNull { it.spoken && StrelkaAlertAlgorithm.isOverspeeding(it.`object`, speedKmh, thresholdKmh) }
    }

    companion object {
        private val nearestFirst = compareBy<State> { it.distanceMeters }.thenBy { it.`object`.id }
    }
}
